pkg/astvalidation/definition_validation.go

```diff
@@ -0,0 +1,42 @@
+package astvalidation
+
+import (
+	"github.com/jensneuse/graphql-go-tools/pkg/ast"
+	"github.com/jensneuse/graphql-go-tools/pkg/astvisitor"
+	"github.com/jensneuse/graphql-go-tools/pkg/operationreport"
+)
+
+func DefaultDefinitionValidator() *DefinitionValidator {
+	validator := &DefinitionValidator{
+		walker: astvisitor.NewWalker(48),
+	}
+
+	validator.RegisterRule(UniqueOperationTypes())
+	validator.RegisterRule(UniqueTypeNames())
+	validator.RegisterRule(UniqueFieldDefinitionNames())
+	validator.RegisterRule(UniqueEnumValueNames())
+	validator.RegisterRule(KnownTypeNames())
+
+	return validator
+}
+
+type DefinitionValidator struct {
+	walker astvisitor.Walker
+}
+
+func (d *DefinitionValidator) RegisterRule(rule Rule) {
+	rule(&d.walker)
+}
+
+func (d *DefinitionValidator) Validate(definition *ast.Document, report *operationreport.Report) ValidationState {
+	if report == nil {
+		report = &operationreport.Report{}
+	}
+
+	d.walker.Walk(definition, definition, report)
```

Comment:
one of args should be nil in this case

Reply:
sadly the walker needs the definition as the second arg for some of the validations, and the first arg is the one on which the walker will walk if I remember right

Reply:
https://github.com/TykTechnologies/graphql-go-tools/blob/master/pkg/introspection/generator.go#L42

For example here, I do not pass second argument

you could try to check will it work in such way
```go
d.walker.Walk(definition, nil, report)
```

Reply:
okay, tried it again and all tests pass - maybe I had some code which needed the definition last time. Now it works like you suggested!